### DISCUSSION TITLE
Serve media files from Postgres instead of Nginx

### DISCUSSION THREAD

**Discussion by yaksheng** [cookiecutter/cookiecutter-django#4642] [Q&A] [answered]
Hi! I refer to cookiecutter-django's [documentation](https://cookiecutter-django.readthedocs.io/en/latest/deployment-with-docker.html#media-files-without-cloud-provider), under "Media files without cloud provider":

"If you chose no cloud provider and Docker, the media files will be served by an nginx service, from a production_django_media volume. Make sure to keep this around to avoid losing any media files."

Media files on my app are user generated and private. I want to store and serve them from the var/lib directory within Postgres after the user is authenticated.

How do I not have the media files served by Nginx? I don't see an option to opt out in the documentation. Thanks!

**Comment by Afrowave** [comment]
Hi @yaksheng,

When you say you want to serve the media files from "var/lib" within Postgres, I would like to understand why you cannot serve them from Nginx and make them accessible only to the user and not to the outside world. That should be easy to do even with regular system level permissions.

Here is a discussion around that from the [Postgres Forum](https://www.postgresql.org/message-id/87psktknwb.fsf@wolfe.cbbrowne.com).

**Reply by yaksheng** [reply]
Hi @Afrowave

If I store and serve private files with Nginx, I would need to handle additional user permissions on Nginx. My attack surface potentially increases too. Also, while Nginx is great for many things including being a media server and reverse proxy, I may move them to hosted blob storage in future.

Given my relative inexperience, I would prefer to manage less systems and reduce complexity. I can further leverage on Postgres' built in row level security to secure these private files.

I would likely use Heroku or Render to host my web service. If so, I would not need Nginx.

Nonetheless, I know my requirements are niche. Thanks for your attention!

**Comment by luzfcb** [comment, marked as answer]
I have never actually used PostgreSQL as a file server, and from my current understanding, I think it was not designed for that and there are probably better alternatives.

If you need to server private files from infrastructure that you maintain yourself (i.e. you are not using Amazon S3 and similar) and limit their access to authenticated users and apply any rules you want, you can use django-downloadview + Nginx X-Accel https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/

https://github.com/jazzband/django-downloadview

**Reply by yaksheng** [reply]
Yes, my (limited) knowledge about Postgres suggests the same too. Thanks for your recommendation!
